### DISCUSSION TITLE
Find workflow definitions based on Variables

### DISCUSSION THREAD

**Discussion by e-mair** [elsa-workflows/elsa-core#5733] [Q&A]
Hello :wave:

I am trying to find workflow definitions based on a number of specifications, one of which, consults a workflow definition's variables.

An example snippet can be found below:
```csharp
public sealed class ManyWorkflowDefinitionsByUserSpecification : Specification<WorkflowDefinition>
{
    /// <summary>
    /// Gets or sets the desired user's ID.
    /// </summary>
    public string Subject { get; init; }

    /// <summary>
    /// Initializes a new instance of the <see cref="ManyWorkflowDefinitionsByUserSpecification"/> class.
    /// </summary>
    /// <param name="subject">The desired user's ID.</param>
    public ManyWorkflowDefinitionsByUserSpecification(string subject)
    {
        Subject = subject;
    }

    /// <inheritdoc/>
    public override Expression<Func<WorkflowDefinition, bool>> ToExpression() => workflowDefinition
        => workflowDefinition.Variables.Has("Users") && workflowDefinition.Variables
            .Get<string[]>("Users")!.Contains(Subject);
}

...

var tenantId = 1337;     // Example tenant ID
var userId = "8d8ced05-bfad-4773-872f-7456e3696d8b";     // Example subject/user ID

var specification = new VersionOptionsSpecification(VersionOptions.Latest)
     .And(new TenantSpecification<WorkflowDefinition>(tenantId))
     .And(new ManyWorkflowDefinitionsByUserSpecification(userId));

var workflowDefinitions = await _workflowDefinitionStore.FindManyAsync(specification, orderBySpecification, paging, cancellationToken);
```
Upon executing our query to MongoDB, I get the following error message:
`System.NotSupportedException: The expression tree is not supported: {document}{Variables}`

As I'm creating this post, I'm wondering if it is at all possible to filter based on a workflow definition's variables as they are persisted as stringified JSON and not actual objects.

Would I have to use MongoDB directly, perhaps with an aggregate pipeline, or is my best choice to query the data without my custom specification, and filter in memory?

**Comment by glime-ai** [comment]
Based on the context provided, it seems you are trying to query workflow definitions by user ID using Elsa Workflows, and you are encountering a `NotSupportedException` when executing the query against MongoDB. The issue arises because the `Variables` property of the `WorkflowDefinition` is likely stored as a JSON string in MongoDB, and the LINQ provider does not support the expression you are trying to use.

Filtering by the `Variables` property directly in the query may not be supported because of the way MongoDB handles JSON data. MongoDB stores JSON documents in a binary-encoded format called BSON, and querying inside a JSON string would require parsing the JSON on the fly, which is not directly supported by LINQ to MongoDB.

To work around this limitation, you have a couple of options:

1. **Filter in Memory**: Retrieve the workflow definitions without the variable filter and then filter the results in memory. This approach may not be efficient if you have a large number of workflow definitions because it involves retrieving more data than necessary from the database.

2. **MongoDB Aggregation Pipeline**: Use MongoDB's aggregation framework to parse and filter the JSON data. This would involve writing a MongoDB query directly, bypassing LINQ, and using MongoDB-specific query capabilities.

3. **Custom Attributes**: If Elsa Workflows supports custom attributes that are searchable, you could store user IDs as custom attributes instead of inside the JSON `Variables` property. This would allow you to query based on these attributes directly.

4. **Database-Specific Features**: Some databases support querying JSON values directly. If MongoDB has such features, you could write a custom query that leverages these features to filter based on the JSON content.

5. **Change Data Model**: If possible, consider changing the data model to store user-related information outside of the JSON `Variables` property in a way that is directly queryable using LINQ.

Given the complexity of the issue, it might be beneficial to consult the Elsa Workflows documentation or community for best practices on querying workflow definitions based on variables. If the feature to query based on JSON content is not currently supported, you might need to consider submitting a feature request or contributing to the project to add such functionality.

For now, if you need to proceed with the current data model, using MongoDB's aggregation pipeline or filtering in memory would be the most immediate solutions.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
